enhancements/storage/csi-efs-operator.md

```diff
@@ -0,0 +1,339 @@
+     At least CI should use this approach to ensure there are no leftovers.
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+Currently, all cloud-based CSI drivers shipped by OCP are installed when a cluster is installed on the particular cloud.
+AWS EFS will not be installed by default. We will include a note about opt-in installation in docs.
+
+After installation of the new operators, either as a fresh install or upgrading the existing one, users must create the
+CR when the operator is installed on a new cluster
+([example of CR for AWS EBS](https://github.com/openshift/cluster-storage-operator/blob/master/assets/csidriveroperators/aws-ebs/10_cr.yaml),
+EFS will be very similar). This is something that was not necessary in the previous releases of
+the operator. We will add a release note and introduce an (info) alert to remind admins to create the CR.
+
+## Design Details
+
+### Rewrite to library-go & ClusterCSIDriver CRD
+
+This should be pretty straightforward. We already ship number of CSI drivers that use library-go. Example usage:
+[AWS EBS CSI driver](https://github.com/openshift/aws-ebs-csi-driver-operator/blob/383a7638b37a4b9a9831c7747e8c499eedcf030f/pkg/operator/starter.go#L67).
+EFS CSI driver operator should not be much more complicated.
+
+Library-go does not support un-installation of CSI driver (and static resources created by `StaticResourcesController`),
+this must be implemented there.
+
+Note: When the *operator* is un-installed, nothing happens to the CRD, CR, the installed CSI driver, existing PVs and
+PVCs. Everything is working, just not managed by the operator. When the *operand* is un-instaled by removing the
+operator CR, the driver will be removed. Existing PVs, PVCs and SharedVolumes will be still present in the API server,
+however, they won't be really usable. Nobody can mount / unmount EFS volumes without the driver.
+
+### `SharedVolume` CRD
+
+Since `SharedVolume` is trivial to handle, just creates PV + PVC, we leave handling of the CRD as a implementation
+detail. We may try to copy code from the community operator, but if it's hard to combine library-go and
+controller-runtime in a single binary, we may as well reimplement the handling to library-go style controller.
+
+### Upgrade from the community operator
+
+Prerequisites:
+* The community aws-efs-operator is installed in OCP 4.8 cluster in `openshift-operators` namespace. The CSI driver runs
+in the same namespace too.
+  * The operator `Subscription` contains the community catalog as the operator source.
+
+Expected workflow (confirmed with OLM team):
+
+1. User upgrades to 4.9. Nothing interesting happens at this point, because the operator `Subscription` still points
+   to the community source. The "old" operator is still running.
+2. User edits the `Subscription` and updates the `source:` to the Red Hat certified catalog.
```

Comment:
IMO these should be entirely separate products. Separate subscriptions etc. As noted above, I think it's acceptable for the upgrade process to be manual; but if it's going to be semi-automated, the new thing can watch & wait for the old version of the driver to disappear before setting up the new one.

Reply:
Reworked the upgrade - user must un-install the old operator + driver and install the new one as a separate step. No automatic upgrade.